source/iterator/snapshot_iterator.go

```diff
@@ -0,0 +1,391 @@
+			if statusResult.NumberOfRecords == 0 {
+				logger.Trace().Msg("Skipping empty export")
+				return false, marketoclient.ErrZeroRecords
+			}
+			return false, nil
+		}
+		return true, nil
+	})
+	if errors.Is(err, marketoclient.ErrZeroRecords) {
+		logger.Trace().Msgf("Skipping,Zero records found for %s", s.exportID)
+		return nil
+	}
+	if err != nil {
+		logger.Err(err).Msg("Error while getting status of export")
+		return err
+	}
+	bytes, err := s.client.FileExportLeads(ctx, s.endpoint, s.exportID)
+	if err != nil {
+		logger.Err(err).Msg("Error while getting file of export")
+		return err
+	}
+	csvReader := csv.NewReader(strings.NewReader(string(*bytes)))
+	_, err = csvReader.Read() // removing the header
+	if err != nil {
+		logger.Err(err).Msg("Error while reading csv")
+		return err
+	}
+	logger.Trace().Msg("Sending csv reader to channel")
+	s.csvReader <- csvReader
+
+	return nil
+}
+
+// prepares and returns record in sdk.Record format. If process fails for any reason, it returns error.
+func (s *SnapshotIterator) prepareRecord(ctx context.Context, data []string) (sdk.Record, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "prepareRecord").Logger()
+	logger.Trace().Msg("Starting the prepareRecord method")
+	var dataMap = marketoclient.GetDataMap(s.fields, data)
+	createdAt, err := time.Parse(time.RFC3339, fmt.Sprintf("%s", dataMap["createdAt"]))
+	if err != nil {
+		logger.Err(err).Msg("Error while parsing createdAt")
+		return sdk.Record{}, fmt.Errorf("error parsing createdAt %w", err)
+	}
+	updatedAt, err := time.Parse(time.RFC3339, fmt.Sprintf("%s", dataMap["updatedAt"]))
+	if err != nil {
+		logger.Err(err).Msg("Error while parsing updatedAt")
+		return sdk.Record{}, fmt.Errorf("error parsing updatedAt %w", err)
+	}
+	if updatedAt.After(s.lastMaxModified) {
+		s.lastMaxModified = updatedAt
+	}
+	position := position.Position{
+		Key:       (dataMap["id"].(string)),
+		CreatedAt: createdAt,
+		UpdatedAt: updatedAt,
+		Type:      position.TypeSnapshot,
+	}
+	pos, err := position.ToRecordPosition()
+	if err != nil {
+		logger.Err(err).Msg("Error while converting position to record position")
+		return sdk.Record{}, fmt.Errorf("error converting position to record position %w", err)
+	}
+	rec := sdk.Record{
+		Payload: sdk.StructuredData(dataMap),
+		Metadata: map[string]string{
+			"id":        position.Key,
+			"createdAt": createdAt.Format(time.RFC3339),
+			"updatedAt": updatedAt.Format(time.RFC3339),
+		},
+		Position: pos,
+		Key: sdk.StructuredData{
+			"id": position.Key,
+		},
+	}
+
+	return rec, nil
+}
+
+// returns Last date from the supplied position.if p is zero value, then it queries least date from the database.
+func (s *SnapshotIterator) getLastProcessedDate(ctx context.Context, p position.Position) (time.Time, error) {
+	logger := sdk.Logger(ctx).With().Str("Method", "getInitialDate").Logger()
+	logger.Trace().Msg("Starting the getInitialDate method")
+	var date = p.CreatedAt.Add(1 * time.Second)
```

Comment:
adding a second to avoid duplicates. if we pull in same time, we may get already read records from marketo. so processing from next second. Also marketo handles data in seconds level only.

Reply:
@hariso any comment on this?

Reply:
> if we pull in same time, we may get already read records from marketo. so processing from next second.

It feels like that's because our query is "return leads where createdAt is greater or equal than lastModified". If so, we need a query like "return leads where createdAt is greater than lastModified".

> Also marketo handles data in seconds level only.

I think that's not a bad reason to to add one second, but what if Marketo changes the precision of timestamps to a millisecond? We may or may not see that soon, but if we can have logic which does not depend on timestamp precision, that would be best.